Mlflow e2e functionality using MySQL backed store.

mparkhe opened the pull request:
- Default artifact location for SQL stores
- Default experiment creation
- Loading SQL store in when using `mlflow ui`
- Unit testing

Commit:
Mlflow e2e functionality using MySQL backed store.

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into sql_e2e

mparkhe commented:
## From UI
![image](https://user-images.githubusercontent.com/6900999/52097031-f3076a80-257d-11e9-98e7-d29753ec522d.png)


## From MySQL 

```
mysql> SELECT
    ->   metrics.run_uuid, metrics.value from metrics
    -> JOIN runs ON metrics.run_uuid = runs.run_uuid
    -> WHERE 1=1
    ->   AND runs.experiment_id = 0
    ->   AND metrics.key = 'm_35'
    ->   AND metrics.value > 0.4 ;
```

| run_uuid                         | value    |
|:----------------------------------|----------:|
| 2b3f979c6c8e4725afd22441705951c5 | 0.590576 |
| 787c2c6b03fc41beb8be06ed583e15b8 | 0.446811 |
| 9e072b7384434f1a8134b0031bdecd5e |  0.76061 |
| b26434bd13d44fb8928a27993d21456c | 0.635022 |

4 rows in set (0.00 sec)

mparkhe commented on mlflow/cli.py:
```diff
@@ -154,7 +154,7 @@ def ui(file_store, host, port, gunicorn_opts):
     # TODO: We eventually want to disable the write path in this version of the server.
     try:
-        _run_server(file_store, file_store, host, port, 1, None, gunicorn_opts)
+        _run_server(file_store, None, host, port, 1, None, gunicorn_opts)
```
In `FileStore` if `artifact_root_uri` is `None`, it uses root directory

```
class FileStore(AbstractStore):
    ....

    def __init__(self, root_directory=None, artifact_root_uri=None):
        """
        Create a new FileStore with the given root directory and a given default artifact root URI.
        """
        super(FileStore, self).__init__()
        self.root_directory = root_directory or _default_root_dir()
        self.artifact_root_uri = artifact_root_uri or self.root_directory
        ...
```

aarondav replied:
Can we add a click alias, and update the documentation to indicate that the available schemas? We can also go ahead and update all public-facing docs to use this new parameter.

aarondav replied:
Same with `server`. There, though, we should be clearer about about what the default is ("in the file store" does not make sense). I think it might be reasonable to make this parameter required when using the sqlalchemy store, at least for the `server` CLI. For the `ui` one, we could reuse `mlruns`, which is the existing default for the file store.

mparkhe commented on tests/store/test_sqlalchemy_store.py:
```diff
@@ -15,18 +15,15 @@
         self.maxDiff = None  # print all differences on assert failures
-        self.store = SqlAlchemyStore(DB_URI)
-        self.engine = sqlalchemy.create_engine(DB_URI)
-        Session = sqlalchemy.orm.sessionmaker(bind=self.engine)
```
`SqlAlchemyStore` already creates `engine` and `session`. Reusing that...

mparkhe commented on mlflow/server/handlers.py:
```diff
@@ -28,8 +28,12 @@ def _get_store():
+        artifact_root = os.environ.get(ARTIFACT_ROOT_ENV_VAR, None)
+        if _is_db_uri(store_dir):
+            from mlflow.store.sqlalchemy_store import SqlAlchemyStore
+            return SqlAlchemyStore(store_dir, artifact_root)
```
reusing `FILE_STORE_ENV_VAR` and `--file-store` option for initialization for now.

aarondav replied:
This one I'm not sure there's a reason to avoid updating this one -- I think this is only used to communicate between `mlflow/server/__init__.py` and `mlflow/server/handlers.py` -- in other words, it is not expected that any user actually sets it (we don't document it).

Commit:
lint fixes

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into sql_e2e

aarondav commented on mlflow/server/handlers.py:
```diff
@@ -17,9 +17,9 @@
     DeleteExperiment, RestoreExperiment, RestoreRun, DeleteRun, UpdateExperiment
 from mlflow.store.artifact_repo import ArtifactRepository
 from mlflow.store.file_store import FileStore
+from mlflow.tracking.utils import _is_db_uri
```
nit: just noticed that "db" is slightly confusing, as there are both database and databricks stores (where the latter is also sometimes shortened to `db_` in mlflow code). Maybe we could change it to `_is_database_uri` to avoid any sort of confusion along these lines.

aarondav commented on mlflow/server/handlers.py:
```diff
@@ -28,8 +28,12 @@ def _get_store():
+            from mlflow.store.sqlalchemy_store import SqlAlchemyStore
+            return SqlAlchemyStore(store_dir, artifact_root)
+        else:
+            _store = FileStore(store_dir, artifact_root)
```
Should we assert the schema is null or `file:` here? It's possible someone tries to put in another kind of store, e.g. a RestStore. That's weird, but it might be good to error explicitly.

aarondav commented on mlflow/server/handlers.py:
```diff
@@ -344,12 +348,6 @@ def _get_artifact_repo(run):
     if run.info.artifact_uri:
         return ArtifactRepository.from_artifact_uri(run.info.artifact_uri, store)
 
-    # TODO(aaron) Remove this once everyone locally only has runs from after
```
Thanks for the cleanup. We should remove the `if` statement (or maybe turn it into an `assert` to make it clear that the run does not have an artifact URI), otherwise we have a path with no `return`.

mparkhe replied:
Good catch. Thanks

aarondav commented on mlflow/store/dbmodels/db_types.py:
```diff
@@ -0,0 +1,11 @@
+SQLITE = 'sqlite'
+MSSQL = 'mssql'
+
+DB_ENGINES = [
```
Could you add a file-level comment indicating what this file contains (namely, the set of sqlalchemy database backend schemas that we support as tracking server backends).

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
+
+    def __enter__(self):
+        if self.db_type == MYSQL:
+            self.session.execute("SELECT @@GLOBAL.sql_mode;")
```
Doesn't this just print the sql_mode? Why is this necessary?

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        if self.db_type == MYSQL:
+            self.session.execute("SELECT @@GLOBAL.sql_mode='';")
```
Shouldn't this be `SET`?

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
+
+class DbHelper:
+    def __init__(self, db_type, session):
+        if db_type not in DB_ENGINES:
```
Given DbHelper is only used during default experiment creation, this seems like a slightly odd place to put this. Why not in the `__init__` of `SqlAlchemyStore`?

mparkhe replied:
This helper is needed for setting `sql_mode` and this setting for different databases. Wanted to avoid adding all these cases in `SqlAlchemyStore.__init__`.

Will rename this to `DatabaseHelper`

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
         db_session = sqlalchemy.orm.sessionmaker(bind=self.engine)
         self.session = db_session()
 
+        if self.artifact_root_uri == DEFAULT_ARTIFACT_ROOT and not exists(DEFAULT_ARTIFACT_ROOT):
```
Why only `if self.artifact_root_uri == DEFAULT_ARTIFACT_ROOT`? It seems reasonable to unconditionally mkdir this (`mkdir` also checks for pre-existence of that file).

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
+            mkdir(DEFAULT_ARTIFACT_ROOT)
+
+        if len(self.list_experiments()) == 0:
+            self._create_default_experiment(db_uri.split(':')[0])
```
For safety, could we use `urlparse(db_uri).scheme`? I don't know if there are any weird shenanigans that could happen, but URIs are complicated.

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -2,27 +2,92 @@
+
+        # Get a list of keys to ensure we have a deterministic ordering
+        columns = list(default_experiment.keys())
+        values = ", ".join([decorate(default_experiment.get(c)) for c in columns])
```
There must be some way to SQL inject here by setting a clever artifact URI...

mparkhe replied:
Let's fix this separately...

aarondav commented on mlflow/tracking/utils.py:
```diff
@@ -169,12 +162,12 @@ def _is_databricks_uri(uri):
 
 def _is_db_uri(uri):
-    if uri.split(':')[0] not in _DBENGINES:
+    if uri.split(':')[0] not in DB_ENGINES:
```
Ah, missed this earlier -- would be great to use `urllib.parse.urlparse(uri).scheme` here too, like for the other `_is_XX_uri`'s

mparkhe replied:
yeah -- i should have caught it too. Thanks for flagging this.

aarondav commented on tests/store/test_sqlalchemy_store.py:
```diff
@@ -37,6 +34,14 @@ def _experiment_factory(self, names):
 
         return self.store.create_experiment(name=names)
 
+    def test_default_experiment(self):
```
Would it be difficult to test that I can delete the default experiment and have it not come back, assuming that I created at least one other experiment?

mparkhe replied:
yes! Was a bit of a pain but wrote a testcase that pins a sqlite file for a test. in the middle of the test, it destroys `SqlAlchemyStore`, and ensures that deleted `Default` experiment stays deleted with new `SqlAlchemyStore` using same sqlite file.

Thanks for the idea...

mparkhe commented on mlflow/store/dbmodels/models.py:
```diff
@@ -97,7 +97,7 @@ class SqlRun(Base):
     source_version = Column(String(50))
     lifecycle_stage = Column(String(20), default=LifecycleStage.ACTIVE)
-    artifact_uri = Column(String(20), default=None)
+    artifact_uri = Column(String(200), default=None)
```
There are a few options for that. The following 2 are widely used with SqlAlchemy.

- SQLAlchemy Migrate
- Alembic

Commit:
refactor to address several review comments
- handle default artifact root paths in CLI and callers
- create experiment transactionality
- constants for sql types and default paths

Commit:
Merge branch 'master' of github.com:mlflow/mlflow into sql_e2e

Commit:
lint fixes

aarondav commented on mlflow/server/handlers.py:
```diff
@@ -27,13 +26,17 @@ def _get_store():
     global _store
     if _store is None:
-        store_dir = os.environ.get(FILE_STORE_ENV_VAR, os.path.abspath("mlruns"))
+        store_dir = os.environ.get(FILE_STORE_ENV_VAR, None)
```
I do think we could rename FILE_STORE_ENV_VAR to BACKEND_STORE_URI_ENV_VAR

aarondav commented on mlflow/store/__init__.py:
```diff
@@ -0,0 +1,5 @@
+Constants shared by several backend store implementations.
+"""
+
+DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH = "./mlruns"
```
Maybe a quick comment on what this is used for (the name is pretty good, but maybe we could contextualize 'default' a bit).

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -1,61 +1,53 @@
+
+    # DB helper methods to allow zero values for columns with auto increments
+    def _set_no_auto_for_zero_values(self):
+        if self.db_type == MYSQL:
```
Btw, were you able to test this with the other supported database engines (postgres/mssql)? Or do we have reason to believe this weirdness only applies to MySQL?

mparkhe replied:
Haven't tried postgreSQL. MariaDB (for example) has a different setting needed for this.

aarondav commented on mlflow/store/sqlalchemy_store.py:
```diff
@@ -120,18 +115,21 @@ def _get_artifact_location(self, experiment_id):
-                self._save_to_db(experiment)
+                eid = new_session.query(SqlExperiment).filter_by(name=name).first().experiment_id
+                experiment.artifact_location = self._get_artifact_location(eid)
+            new_session.commit()
```
This is magic! Cool!

mparkhe replied:
Your idea! Weird that it works ... but then again, using local session for this was needed.

Commit:
Addressing review comments

Commit:
fixing broken test
